Framework: CSS: Add stylizer component

scruffian opened the pull request:
This PR is the next part of https://github.com/Automattic/wp-calypso/issues/11959, continuing the work done in https://github.com/Automattic/wp-calypso/issues/12451. It adds a new higher order component, called `Stylizer`. This outputs the CSS for a component when the component is mounted, so that we only load the CSS when it is needed.

I have implemented this for one component - `VerticalNav` as a demonstration of how it will work, and that it does work.

We use Webpack to load all required dependencies and wrap with Stylizer. There are the following items which need to be resolved in the follow up PRs before we start to roll out it on a wider scale:

* We should import style.scss file only when it's present in the folder where React component is located
* RTL support is missing
* SSR support
* Documentation
  
#### Testing instructions
  
1. Run `git checkout add/stylizer` and start your server, or open a [live branch](https://delphin.live/?branch=add/stylizer)
2. Open the [`Domain management` page](http://calypso.localhost:3000/domains/manage/example.com/edit/example.com)
3. Check that the vertical nav CSS is loaded in the head:
<img width="617" alt="screen shot 2017-04-27 at 10 15 37" src="https://cloud.githubusercontent.com/assets/275961/25476767/2483ee4a-2b33-11e7-98f7-bfe435c7eb7f.png">
4. Navigate to the reader, and check that the CSS is removed from the head.

#### Reviews
  
- [ ] Code
- [ ] Product
   
@Automattic/sdev-feed

folletto commented:
How does this impact server-side rendering?

gziolo commented on client/components/vertical-nav/index.jsx:
```diff
@@ -2,6 +2,12 @@
  * External dependencies
  */
 import React from 'react';
+import withStyles from 'isomorphic-style-loader/lib/withStyles';
```
I think we shouldn't merge changes in this file. Can you split this to its own file so we could merge only code that  enabled stylizer?

scruffian replied:
Its easy to remove, but why?

gziolo replied:
We agreed to split CSS Modules work into smaller PRs. It doesn't seem like we should enable styles before we finish all other required tasks.

nb replied:
It was a good example what a component would look like, so it was useful to me.

gziolo commented:
> Your build ran 6867 tests with 1 failure
> Root Suite my-sites upgrades domain-management edit mapped-domain "before all" hook - "before all" hook
> moreError: Cannot find module './style'

`./style` files need to be mocked out when a component is loaded when running unit tests.

scruffian commented:
> ./style files need to be mocked out when a component is loaded when running unit tests.

Thanks. How?

scruffian commented:
> How does this impact server-side rendering?

@folletto This PR will have no impact on SSR. We'll need to think about how to address that in a future PR.

blowery commented on server/pages/index.js:
```diff
@@ -143,7 +143,13 @@ function getDefaultContext( request ) {
 		devDocsURL: '/devdocs',
-		store: createReduxStore( initialServerState )
+		store: createReduxStore( initialServerState ),
+		css: new Set(),
```
a set doesn't maintain order, while CSS can be order dependent. Does that matter, or are we assuming that our CSS is not order dependent?

gziolo replied:
It would be a very bad assumption :) Our CSS depends in a few places on the order loading.

Tug replied:
Should we update to an array then?

jsnajdr replied:
`Set.prototype.forEach` respects the original insertion order, [the spec says](https://www.ecma-international.org/ecma-262/6.0/#sec-set.prototype.foreach). I'm not sure if the iterator (e.g., `set.entries()`) does, too -- the spec is not so clear about that.

ehg commented:
Do we really have to add another HoC to each component? I'm concerned this will get messy.

gziolo commented on client/components/vertical-nav/index.jsx:
```diff
@@ -13,4 +19,4 @@ const VerticalNav = React.createClass( {
 } );
 
-export default VerticalNav;
+export default withStyles( style )( VerticalNav );
```
Could we omit `import style from './style';` line and use the following instead:

```js
export default withStyles( './style' )( VerticalNav );
```
:) ?

gziolo commented:
@ehg you can compare it with Styled Components: https://github.com/styled-components/styled-components#usage or Glamorous https://github.com/paypal/glamorous#this-solution. They use components which are wrappers, and those two don't work with external SCSS files.

I think it's a reasonable trade-off. We can try to make this HOC easier to use and promote composing HOCs :)

gziolo commented:
@scruffian a94bb3fcad49b8bbd1780c0e54949a69b92f2f4e fixes failing test, but I don't think this is a solution that would scale. Is it possible that there are no tests in Delphin that would load Component importing styles? Did you have similar issues when working on get.blog?

scruffian commented:
@gziolo I don't think we have any component tests in Delphin.

scruffian commented:
> Do we really have to add another HoC to each component? I'm concerned this will get messy.

@ehg do you have another suggestion for how to achieve this?

ehg commented:
> @ehg do you have another suggestion for how to achieve this?

@scruffian — I think we discussed some alternatives in two of our framework hangouts. If we want to take this approach, we could possibly write a babel plugin to add the HoC. I see `withStyles` as very boiler-plate, of which we have enough in Redux :)

The alternative is writing our own webpack loader, and inspecting the filesystem graph and possibly `sections.js` — instead of the JS module graph. Which, IMHO is more natural, although an unknown in terms of work.

gziolo commented:
@ehg I did a quick research and I found this: http://stackoverflow.com/questions/32138145/optional-dependencies-in-webpack/32164232#32164232. Is it similar to what you propose? It misses the part where you inject CSS into DOM, but otherwise looks quite promising :)

gziolo commented:
There is also https://github.com/deepsweet/baggage-loader which seems to be tailored to this use case!

stephanethomas commented:
@ehg: I see your point for using a loader or a plugin to avoid that boilerplate code. However I'm concerned that by using this sort of magic we will make people that are not familiar with this pattern wonder how this could possibly work. We can use [`compose()`](https://github.com/Automattic/wp-calypso/blob/d53671ef1a6b1e023531d469a62a13a390ac2377/client/me/account-password/index.jsx#L186-L193) to make code using HOCs a bit more readable. Finally, I also suspect that having a loader transforms our code will increase our build time.

nb commented:
I also feel strongly against another wrapper higher-order component. Almost every component will need CSS, so we will have to add it to almost all of our component files and that’s too much.

“Magic” has its place and connecting components and CSS files sounds like a great use-case – saves a ton of boilerplate, easy to explain (“all CSS files in the component directory are loaded alongside the component”), and very close to the way we have been working now.

Currently in order to know that all component SCSS files are referenced from [a single place](https://github.com/Automattic/wp-calypso/blob/c82b35a46d4d98ae274f04407352baef44b5de9c/assets/stylesheets/_components.scss) we would either need to look at how the build works or we would need to read the documentation. If we implicitly load the CSS files in the directory of a component, the developer experience wouldn’t be much worse, if any. Reasoning about the system would be as easy. And boilerplate would be 0.

coderkevin commented:
>Almost every component will need CSS

I tend to write several container components that don't use CSS or i18n. They only output other React components.

But this leads me to wonder: What if we could combine `localize()` with this other HOC? Chances are you'd need one if you need the other. Could we have a single HOC for all Presentational components?

yurynix commented:
Regarding the Webpack magic suggested here, as most "magics" go, for DEVs unfamiliar with it, it would be hard to reason about, while using an explicit HoC will set them on the right direction.

Lets consider a few cases we currently have:
https://github.com/Automattic/wp-calypso/tree/master/client/reader/tag-stream
Which one of those components should include `style.scss` ?
`index.js` which has no JSX / React ?
`controller.js` which doesn't export a react component?
one of the `.jsxs` which one? all of them?

https://github.com/Automattic/wp-calypso/tree/master/client/components/seo
This structure have the `style.scss` outside of each component dir, our webpack plugin will have a hard time auto-guessing that.
( we should consider splitting that `style.scss` for each component's dir )

We might want to re-consider magic Webpack compilation in the future, but the first step forward IMO should be plain magicless HoC.

blowery commented:
> I also feel strongly against another wrapper higher-order component. Almost every component will need CSS, so we will have to add it to almost all of our component files and that’s too much.

scruffian commented:
I added a commit to remove the higher order component and use Webpack to load `withStyles`. This approach requires us to specify individual components in the Webpack config until we have converted all the components to work with this approach. I think that's a task for a different PR.

gziolo commented on client/components/vertical-nav/item/style.scss:
```diff
@@ -1,3 +1,6 @@
+@import 'assets/stylesheets/shared/_colors';
```
I was wondering if it would be possible to import one file that contains all variables, functions, mixins and stuff that doesn't produce additional CSS itself. With that in place we could use another Webpack loader to prepend it without any changes to the existing files :)

gziolo replied:
After some discussion we decided to take approach of adding explicit imports. It looks like we can use codemod provided by Yury to make it work.

drewblaisdell commented on webpack.config.js:
```diff
@@ -65,11 +65,27 @@ const webpackConfig = {
+				]
+			},
+			{
+				test: /\/client\/components\/vertical-nav.+index\.jsx$/,
```
Are there any other changes that need to be made before we can switch this to `^\/client/(.)*\/index.jsx$`?

gziolo replied:
* sass imports are missing in *.sass files (variables, mixins, functions)
* we need to find a way to conditionally load style.scss - don't try to load the file if it is missing
* we need to cover the case where we override index.jsx with a different name using `package.json`
* everything else that is going to pop up :)

drewblaisdell replied:
The third point sounds like something we can merge v1 of this without, yeah?

gziolo replied:
Yes. We could also get rid of those package.json files to fix missed files :)

ockham commented:
> But this leads me to wonder: What if we could combine `localize()` with this other HOC?

@coderkevin I'd rather keep concerns separate. 

One question that comes to mind -- if we want i18n to be smarter (e.g. I've heard that right now, site settings do a full page reload upon locale change cc @jsnajdr), can we still use a webpack loader/babel plugin? Or does it need to be more dynamic than that? _Smarter_ could mean _better integrated with Redux_. (I'll add SSR as another use case that works best with state contained in our Redux tree.) I think that we're treating i18n state quite implicitly right now; it's essentially stored inside the `i18n-calypso` module I think.

So that's something that we should maybe tackle first before turning it into webpack magic, since we might need to e.g. add a locale argument from Redux to `localize`.

---

Another option that comes to mind: `class CalypsoBaseComponent extends React.Component {...}`.

jsnajdr commented:
@ockham I outlined the steps necessary to remove the reload on locale change in #13712. It turns out that the `localize` HOC is already smart enough -- it listens for a change of locale in the `i18n` module and forces an update in such case. If the `localize` HOC keeps that capability, we are ok.

There are problems with components that don't use neither the HOC or the mixin and just call the `i18n.translate` function directly. Also, there are two separate CSS bundles for LTR and RTL languages -- `style.css` and `style-rtl.css`. That's solvable, too.

gziolo commented:
@ehg @nb we have a proposal which is boilerplate free. We use Webpack to load all required dependencies and wrap with Stylizer. There are the following items which need to be resolved before we start to roll out it on a wider scale:

* We should import style.scss file only when it's present in the folder where React component is located
* RTL support is missing

gziolo commented on assets/stylesheets/shared/_component.scss:
```diff
@@ -0,0 +1,6 @@
+@import 'colors';                       // import all of our wpcom colors
+@import 'typography';                   // all the typographic rules, variables, etc.
+@import 'mixins/mixins';                // sass mixins for gradients, bordius radii, etc.
+@import 'extends';                      // sass extends for commonly used styles
```
We probably need to move more code here. We will discover what we missed when we start playing with more folders.

yurynix replied:
Another [option](https://github.com/Automattic/wp-calypso/blob/05f3968e3b45be9270479db618b4bd464c48ab58/bin/refactor-to-import-sass.js#L566-L656) is add the missing dependencies where they should've been - the module's `style.scss` as a one-time codemod.

gziolo replied:
Yes, we are going to try the one-time codemod approach first.

gziolo commented on package.json:
```diff
@@ -134,6 +137,7 @@
     "valid-url": "1.0.9",
     "walk": "2.3.4",
     "webpack": "1.13.1",
+    "webpack-append": "0.1.2",
```
`webpack-append` is so simple that I we might want to rewrite it ourselves :)

BTW, it should be called prepend!

yurynix commented on webpack.config.js:
```diff
@@ -65,11 +65,28 @@ const webpackConfig = {
+				]
+			},
+			{
+				test: /\/client\/components\/vertical-nav.+index\.jsx$/,
```
I'm not sure how will that work in the "general" case, should we create a 'generic' loader
that will do that instead of hard-coding it?

```js
{
   test: /\.scss/,
   loaders: [
'isomorphic-style-loader',
					'css-loader?camelCase=dashes&importLoaders=1',
 					'sass-loader',
 					"maybe-append-import-and-hoc",
   ]
```
? Or i missed something?

gziolo replied:
Yes, we limited it only to one file to validate it works properly. We plan to provide general solution in the follow up PR.

As mentioned in a different place, we still need to cover the following items:
* We should import style.scss file only when it's present in the folder where React component is located
* RTL support is missing
* SSR support

yurynix commented:
Thank you @gziolo and @scruffian for pushing that forward!

1. I preferred the explicit solution where we do a codemod to add the HoC and the import; however that's also a good solution.
2. I think the code is good and we can move forward ; I'm not sure if we really need to hardcode `/\/client\/components\/vertical-nav.+index\.jsx$/` since we're going to remove it anyway.

LGTM :shipit:

aduth commented on server/pages/index.jade:
```diff
@@ -54,6 +54,8 @@ html(lang=lang, dir=isRTL ? 'rtl' : 'ltr', class=isFluidWidth ? 'is-fluid-width'
 			else
 				link(rel='stylesheet', href=urls['style.css'])
+		style(type='text/css' id='critical-path-css')
+			!= Array.from(css).join('')
```
How do we decide what is critical path CSS? From what I can tell, this is including every component's styles (currently just VerticalNav)?

Minor: Spacing between parentheses.

Tug replied:
Currently, we are including the styles of all **mounted** components which uses the `withStyles` HOC.

The way this HOC works is that when we render a page on the server, `insertCss` is called [for each component](https://github.com/kriasoft/isomorphic-style-loader/blob/master/src/withStyles.js#L22) and the styles ([`as text`](https://github.com/kriasoft/isomorphic-style-loader/blob/master/src/index.js#L30)) are added to the `css` Set.
Once all component are rendered we have a "critical path CSS" as a set of styles.

We can always improve it by customizing the `insertCss` function later, but I think the default behavior is already pretty good.

Tug replied:
Btw @scruffian we should probably remove those until we have SSR working, what do you think?

scruffian replied:
@tug I think this change will be needed for SSR so we might as well make it now. We can always revisit it if we spot other issues down the line.

Tug replied:
I meant, we should do it when we implement SSR since it's only really needed in that case. Ok to keep it now if you prefer though

aduth commented on webpack.config.js:
```diff
@@ -65,11 +65,28 @@ const webpackConfig = {
 	},
 	resolve: {
-		extensions: [ '', '.json', '.js', '.jsx' ],
+		extensions: [ '', '.json', '.js', '.jsx', '.scss' ],
```
Do we need to resolve the extension if we're not expecting the SASS files to be imported by developers? IIRC, this is the difference between allowing:

```js
import './style.scss';
```

And:

```js
import './style';
```

(Latter with `extensions` defined)

gziolo replied:
Yes, makes sense. We don't need this change anymore.

gziolo replied:
Removed

aduth commented on webpack.config.js:
```diff
@@ -65,11 +65,28 @@ const webpackConfig = {
+					'isomorphic-style-loader',
+					'css-loader?camelCase=dashes&importLoaders=1',
+					'sass-loader',
+					"webpack-append?@import 'assets/stylesheets/shared/component';",
```
You can do this without an additional loader using `sass-loader`'s `data` query argument.

Example: https://github.com/WordPress/gutenberg/blob/935b958/webpack.config.js#L60

gziolo replied:
Nice one, this is exactly what we need. Thanks for hint!

gziolo replied:
Updated

Tug commented on client/lib/react-helpers/index.js:
```diff
@@ -35,7 +36,11 @@ export function renderWithReduxStore( reactElement, domContainer, reduxStore ) {
 	return ReactDom.render(
-		React.createElement( ReduxProvider, { store: reduxStore }, reactElement ),
+		<ReduxProvider store={ reduxStore }>
+			<Stylizer onInsertCss={ insertCss }>
```
I would move the `Stylizer` to `renderPage` in principle so `renderWithReduxStore` does what it says it does. But I think we don't use `renderPage` everywhere yet so that's for another PR.

gziolo replied:
Do we need it for every page or can we put it in the top component like `<Layoyt />`?

nb replied:
I don't think it makes a serious functional difference, but in `Layout` it would be a bit easier to discover.

gziolo replied:
It's just easier to put it in `<Layout />`, we have only 2 versions.

scruffian replied:
> But I think we don't use renderPage everywhere yet so that's for another PR.
Agreed

ehg commented:
I think this is a step in the right direction, although a rough plan for RTL and SSR would be great.

@jsnajdr I seem to remember you having some ideas on combined RTL bundles, would you care to chime in? :)

jsnajdr commented:
> I seem to remember you having some ideas on combined RTL bundles, would you care to chime in?

I researched the current RTL workflow in #13712. We now use the RTLCSS tool to transform the `style.css` bundle into a `style-rtl.css` and load only one of them depending on the interface language. The transformation replaces properties like `margin-left` into `margin-right` and vice versa.

My proposal is to unify both CSS bundles into one, with a postprocessing step that transforms
```css
.button {
  margin-left: 10px;
}
```
into
```css
[dir='ltr'] .button {
  margin-left: 10px;
}
[dir='rtl'] .button {
  margin-right: 10px;
}
```
That can be achieved by either patching the RTLCSS tool, or use it as a starting point for developing our own PostCSS plugin.

I'm not very familiar with the CSS modules and CSS-in-JS tools, so I'm not sure if plugging in a custom PostCSS transformer into the style loading pipeline is viable or not.

yurynix commented:
@jsnajdr Could you elaborate on the reasoning behind combining it to a single bundle? Is it just to avoid reload? 
While that is a great benefit, the tradeoff is increasing the bundle size for users who might not ever use RTL, is it "worth" it?

gziolo commented:
@jsnajdr I think updating language is an edge case, which happens on a very rare occasion. I think most of the users don't change it at all. Unfortunately we don't track how often this setting gets updated.

Let me ask another question, what's wrong with having the page reload when user switches to a different language?

nb commented on webpack.config.js:
```diff
@@ -65,7 +65,23 @@ const webpackConfig = {
+				test: /\/client\/components\/vertical-nav.+index\.jsx$/,
+				loaders: [
+					'imports?style=./style.scss,withStyles=isomorphic-style-loader/lib/withStyles',
+					'exports?default=withStyles.default( style )( exports.default )'
```
How would this work if we want to do it on all components? Are all of our components in `.jsx` files? Do we always have only one component per module?

Can we find a smarter way to do this? What were the other ideas?

gziolo replied:
Very good point. I think we need to validate the following question:

> Is it possible to create Webpack loader which scans every `*.js` and `*.jsx` files and picks only those that use React. For all those files it should also look if there is a `style.scss` file in the same folder. If all those requirements are met for a single file it should modify this file as above.

Does it make sense? I think it would cover most of the use cases. Unless we have files which use React but don't expose components ...

yurynix replied:
Another problem is, that in a single folder there might be many `.jsx?` components, but only the "entry point" component should have the `import`.

scruffian replied:
Can we just do it for `index.jsx` components when there is a `style.scss` file present?

gziolo replied:
We would like to verify if we can use `index.jsx` plus `style.scss` pair first. It seems like the most common case where we should auto-import styles.

nb commented:
Thanks for continuing to work on that :)

The current version looks good on the devex front – no changes in components needed. The main question is, even if we merge this PR and it works well, how do we scale that in terms of configuration and matching components to stylesheets. Essentially the first item in your TODO list above.

Also, has anybody looked at the implications of https://twitter.com/slightlylate/status/845435543647248384? For super small stylesheets it may make sense to ship them inline, but for bigger ones we both can look at downloading them separately and we may avoid double parsing/extra DOM mutations.

jsnajdr commented:
@gziolo Avoiding the page reload after language change is one of ideas I explored during my trial. I agree that the benefit for the user experience is not that big -- it's a rare action and one reload more or less is not a big deal.

More important motivation is removing some complexity from the `me/account` form. The reload is a side-effect that doesn't fit very well into the Redux way. There is a special case for saving the `language` value vs saving everything else. There is an elaborate process to schedule displaying of a success notice after save and reload. I remember that @ockham was very in favor of removing the reload every time I mentioned it on Slack :)

@yurynix Regarding the bundle size increase, the diff between `style.css` and `style-rtl.css` is 250 kB big. Since the whole bundle is 1.45 MB, the bundle would get approx 20% bigger. This is a rough estimate -- maybe we can invent some optimizations, maybe not.

gziolo commented:
It looks like Webpack team has interesting plans in regards to code splitting CSS: https://medium.com/webpack/the-new-css-workflow-step-1-79583bd107d7.

gziolo commented:
We are putting this on hold, because we didn’t come up with a satisfying solution how to move CSS out of JS bundles. We also discussed that RTL transformations should be done as part of the build process, not at the runtime. We will share more detailed information in the coming up post.

In the meantime we probably need to split our big CSS file into smaller files to increase the probability that user is going to take advantage of browser caching.
